File: ios/UIComponent.m
#import "UIComponent.h"
#import "ReactNativeBrownfield.h"
#import "ReactNativeComponentFactory.h"

@implementation UIComponent

@synthesize reactComponent = _reactComponent;
@synthesize hasChildren = _hasChildren;

-(id)initWithFrame:(CGRect)frame {
    _hasChildren = NO;
    if([self class] == [UIComponent class]) {
        [self doesNotRecognizeSelector:_cmd];
        return nil;
    }
    
    if ((self = [super initWithFrame:frame])){
         [self addReactSubview];
    }
    return self;
}

-(id)initWithCoder:(NSCoder*)aDecoder {
    _hasChildren = NO;
    if([self class] == [UIComponent class]) {
        [self doesNotRecognizeSelector:_cmd];
        return nil;
    }
    
    if ((self = [super initWithCoder:aDecoder])){
         [self addReactSubview];
    }
    return self;
}

-(void)layoutSubviews {
    [self updateProps];
    [super layoutSubviews];
}

-(void)addSubview:(UIView *)view {
    [[ReactNativeBrownfield shared] registerView:_reactComponent.uuid view:view];
    _hasChildren = YES;
    [self updateProps];
}

-(void)initializeReactSubview:(UIView *)view {
    [super addSubview:view];
}

-(void)removeSubviews {
    [[ReactNativeBrownfield shared] unregisterView:_reactComponent.uuid];
    _hasChildren = NO;
    [self updateProps];
}

-(NSDictionary*)getProps {
    NSMutableDictionary *props = [NSMutableDictionary new];

    props[@"uuid"] = _reactComponent.uuid;
    if (_hasChildren) {
        props[@"hasChildren"] = @YES;
    }
    
    return props;
}

-(void)updateProps {
    dispatch_async(dispatch_get_main_queue(), ^{
        NSMutableDictionary *props = [[self getProps] mutableCopy];

        if (self->_hasChildren) {
            props[@"hasChildren"] = @YES;
        }

        self->_reactComponent.appProperties = props;
    });
}

-(void)addReactSubview {}

@end

File: android/src/main/java/com/callstack/reactnativebrownfield/UIComponent.kt
package com.callstack.reactnativebrownfield

import android.content.Context
import android.os.Handler
import android.util.AttributeSet
import android.view.View
import android.widget.FrameLayout


abstract class UIComponent: FrameLayout {
    constructor(context: Context): super(context)
    constructor(context: Context, attrs: AttributeSet?): super(context, attrs)
    constructor(context: Context, attrs: AttributeSet?, defStyleAttr: Int): super(context, attrs, defStyleAttr)

    var reactComponent: ReactNativeComponent? = null
    var hasChildren = false

    protected abstract fun addReactSubview()
    protected open fun getProps() : HashMap<String, Any> {
        val map = hashMapOf<String, Any>("hasChildren" to hasChildren)

        if (reactComponent != null) {
            map["uuid"] = reactComponent!!.uuid
        }

        return map
    }

    private fun updateProps() {
        val mainHandler = Handler(context.mainLooper)

        val map = getProps()

        mainHandler.post {
            reactComponent?.appProperties = PropsBundle.fromHashMap(map)
        }
    }

    protected fun initializeReactSubview() {
        super.addView(reactComponent)
    }

    override fun addView(child: View?) {
        if (child != null && reactComponent != null) {
            hasChildren = true
            updateProps()
            ReactNativeBrownfield.shared.registerView(reactComponent!!.uuid, child)
        }
    }

    override fun removeAllViews() {
        super.removeAllViews()
        if (reactComponent != null) {
            hasChildren = false
            updateProps()
            ReactNativeBrownfield.shared.unregisterView(reactComponent!!.uuid)
        }

    }

    override fun onLayout(changed: Boolean, left: Int, top: Int, right: Int, bottom: Int) {
        super.onLayout(changed, left, top, right, bottom)
        updateProps()
    }
}